pkg/modifycontroller/controller.go

```diff
@@ -0,0 +1,378 @@
+	pvcRateLimiter workqueue.RateLimiter) ModifyController {
+	pvInformer := informerFactory.Core().V1().PersistentVolumes()
+	pvcInformer := informerFactory.Core().V1().PersistentVolumeClaims()
+	vacInformer := informerFactory.Storage().V1alpha1().VolumeAttributesClasses()
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartStructuredLogging(0)
+	eventBroadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: kubeClient.CoreV1().Events(v1.NamespaceAll)})
+	eventRecorder := eventBroadcaster.NewRecorder(scheme.Scheme,
+		v1.EventSource{Component: fmt.Sprintf("external-resizer %s", name)})
+
+	claimQueue := workqueue.NewNamedRateLimitingQueue(
+		pvcRateLimiter, fmt.Sprintf("%s-pvc", name))
+
+	ctrl := &modifyController{
+		name:                    name,
+		modifier:                modifier,
+		kubeClient:              kubeClient,
+		pvSynced:                pvInformer.Informer().HasSynced,
+		pvcSynced:               pvcInformer.Informer().HasSynced,
+		vacSynced:               vacInformer.Informer().HasSynced,
+		claimQueue:              claimQueue,
+		volumes:                 pvInformer.Informer().GetStore(),
+		claims:                  pvcInformer.Informer().GetStore(),
+		volumeAttributesClasses: vacInformer.Informer().GetStore(),
+		eventRecorder:           eventRecorder,
+	}
+
+	// Add a resync period as the PVC's request size can be resized again when we handling
+	// a previous resizing request of the same PVC.
+	pvcInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ctrl.addPVC,
+		UpdateFunc: ctrl.updatePVC,
+		DeleteFunc: ctrl.deletePVC,
+	}, resyncPeriod)
+
+	return ctrl
+}
+
+func (ctrl *modifyController) initUncertainPVCs() error {
+	ctrl.uncertainPVCs = make(map[string]v1.PersistentVolumeClaim)
+	for _, pvcObj := range ctrl.claims.List() {
+		pvc := pvcObj.(*v1.PersistentVolumeClaim)
+		if pvc.Status.ModifyVolumeStatus != nil && (pvc.Status.ModifyVolumeStatus.Status == v1.PersistentVolumeClaimModifyVolumeInProgress || pvc.Status.ModifyVolumeStatus.Status == v1.PersistentVolumeClaimModifyVolumeInfeasible) {
+			pvcKey, err := cache.MetaNamespaceKeyFunc(pvc)
+			if err != nil {
+				return err
+			}
+			ctrl.uncertainPVCs[pvcKey] = *pvc.DeepCopy()
+		}
+	}
+
+	return nil
+}
+
+func (ctrl *modifyController) addPVC(obj interface{}) {
+	objKey, err := getObjectKey(obj)
+	if err != nil {
+		return
+	}
+	ctrl.claimQueue.Add(objKey)
+}
+
+func (ctrl *modifyController) updatePVC(oldObj, newObj interface{}) {
+	oldPVC, ok := oldObj.(*v1.PersistentVolumeClaim)
+	if !ok || oldPVC == nil {
+		return
+	}
+
+	newPVC, ok := newObj.(*v1.PersistentVolumeClaim)
+	if !ok || newPVC == nil {
+		return
+	}
+
+	// Only trigger modify volume if the following conditions are met
+	// 1. Feature gate is enabled
+	// 2. Non empty vac name
+	// 3. oldVacName != newVacName
+	// 4. PVC is in Bound state
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeAttributesClass) {
```

Comment:
the controller is responsible for moditying a volume, so I think it is better to move the feature-gate checking into main.go. If the feature-gate is off or vac API is unvaliable, the controller wouldn't be initiated and started. In this case, only resizer controller is started.